### ISSUE TITLE
How many docker swarm managers?

### ISSUE THREAD

**Issue by manast** [docker/docker#26919]
I failed to find in the documentation or in the issues any information regarding how many swarm managers I should declare. Sure, more than one if you want high availability, but how many? and why not make all nodes managers? what is the tradeoff of being a manager vs a worker?

**Comment by cpuguy83** [comment]
The more managers, the slower things will be.
Swarm is using raft for consensus, which requires a majority of participants to agree a particular change.
If there are more managers, it will take longer to reach consensus.
But with more managers comes greater fault tolerance.
In addition, it may not always be worthwhile (and potentially actively harmful) to add managers.

Swarm uses raft for consensus, which requires a majority of managers to agree on a change.

- With 1 manager node, you must always have the manager up
- With 2 manager nodes, both managers are required for quorum
- With 3 managers, only 2 nodes are required for quorum

Here is a chart to explain this further:

|# of managers| failure tolerance      | required for quorum |
|---------------|--------------------|---------------------|
|        1                |             0                  |            1
|        2                |             0                  |            2
|        3                |             1                  |            2
|        4                |             1                  |            3
|        5                |             2                  |            3
|        6                |             2                  |            4
|        7                |             3                  |            4

Basically, you never want an even number of nodes, this is wasteful.
You probably would never need more than 7 managers, and probably 5 is more than enough.
But this will be up to your environment.

This will be the same advice for running etcd and consul.

**Comment by thaJeztah** [comment]
Documentation for this is here; https://docs.docker.com/engine/swarm/admin_guide/#/add-manager-nodes-for-fault-tolerance, so I think we can close this issue

**Comment by manast** [comment]
@cpuguy83 thanks for the info. 
@thaJeztah I am not really sure that the information provided by @cpuguy83 is available at your link, or at least not so complete.

**Comment by kurtismullins** [comment]
As this popped up high on my Google search and am still left with questions, I referenced this ticket on StackOverFlow:

https://stackoverflow.com/questions/48853473/pros-and-cons-of-running-all-docker-swarm-nodes-as-managers
